app/waf/src/controllers/application.controller.ts

```diff
@@ -299,96 +437,50 @@ export class ApplicationController {
     responses: {
       '204': Schema.emptyResponse('Successfully deploy Application resource'),
       '404': Schema.notFound('Can not find Application resource'),
+      '422': Schema.unprocessableEntity('Fail to deploy Application resource'),
     },
   })
   async deployById(
     @param(Schema.pathParameter('applicationId', 'Application resource ID'))
     id: string,
-  ): Promise<Object> {
+  ): Promise<string> {
     let application = await this.applicationRepository.findById(id);
 
-    let services = await this.applicationRepository
-      .services(application.id)
-      .find();
-    if (services.length === 0) {
+    if (!application.adcId || !application.currentDeclarationId) {
       throw new HttpErrors.UnprocessableEntity(
-        'No service in Application ' + application.id,
+        'No target ADC or no Application declaration to perform deploy action',
       );
     }
 
-    let params: {[key: string]: Object} = {
-      application: application,
-    };
-
-    let tenantAssocs = await this.tenantAssociationRepository.find({
-      limit: 1,
-      where: {
-        tenantId: application.tenantId,
-      },
-    });
-
-    if (tenantAssocs.length === 0) {
-      tenantAssocs = await this.tenantAssociationRepository.find({
-        limit: 1,
-        where: {
-          tenantId: 'default',
-        },
-      });
+    let adc = await this.adcRepository.findById(application.adcId);
+    let declaration = await this.declarationRepository.findById(
+      application.currentDeclarationId,
+    );
 
-      if (tenantAssocs.length === 0) {
-        throw new HttpErrors.UnprocessableEntity(
-          'No ADC associated with Application ' + application.id,
-        );
-      }
-    }
+    let req = new AS3DeployRequest(adc, application, declaration);
+    return await this.as3Service.deploy(AS3_HOST, AS3_PORT, req);
+  }
 
-    let adcs = await this.adcRepository.find({
-      where: {
-        id: tenantAssocs[0].adcId,
-      },
-    });
+  @post(prefix + '/applications/{applicationId}/cleanup', {
```

Comment:
Move to declaration.controller. 

Or it makes confusion:

@post(prefix + '/applications/{applicationId}/cleanup'
is too easy to be thought of 
@del(prefix + '/aplications/{applicationId}').

Reply:
Maybe we can give a better name to this action.

Reply:
I prefer to move declarations REST APIs to /declarations (declaration.controller.ts).

Reply:
Done